File: language_construct_performance/php/scripts/update_readme.php
<?php

declare(strict_types=1);

function formatMB(int $mb): string
{
    $gb = number_format($mb / 1024, 1);
    return sprintf("%s MB (~%s GB)", number_format($mb), $gb);
}

function formatMS(int $ms): string
{
    return sprintf("%s ms", number_format($ms));
}

function run()
{
    $readmePath = dirname(__DIR__) . '/README.md';
    $statsPath = dirname(__DIR__) . '/data/stats.json';
    $statsVarPath = dirname(__DIR__) . '/data/stats_variable.json';

    $stats = null;
    $statsVar = null;

    if (file_exists($statsPath)) {
        try {
            $content = file_get_contents($statsPath);
            if ($content !== false) {
                $stats = json_decode($content, true);
            }
        } catch (\Throwable $e) {
            echo "Error parsing {$statsPath}: " . $e->getMessage() . "\n";
        }
    }

    if (file_exists($statsVarPath)) {
        try {
            $content = file_get_contents($statsVarPath);
            if ($content !== false) {
                $statsVar = json_decode($content, true);
            }
        } catch (\Throwable $e) {
            echo "Error parsing {$statsVarPath}: " . $e->getMessage() . "\n";
        }
    }

    if (!$stats && !$statsVar) {
        echo "No valid stats files found in data/ directory!\n";
        exit(1);
    }

    $numEntries = $stats['numEntries'] ?? $statsVar['numEntries'] ?? 1000000;
    $formattedEntries = number_format($numEntries);

    $newSection = "<!-- BENCHMARK_RESULTS_START -->\n";
    $newSection .= "## Benchmark Results ({$formattedEntries} Entries)\n\n";
    $newSection .= "Here are the actual measured results from running the isolated benchmark suite under PHP with **{$formattedEntries} entries**:\n";

    if ($stats) {
        $newSection .= "\n### 1. Fixed Properties (Uniform Structural Shape)\n\n";
        $newSection .= "| Metric | Plain Object (Array) | Value Object | Value Object Minimal |\n";
        $newSection .= "| :--- | :---: | :---: | :---: | :---: |\n";
        $newSection .= sprintf("| **Creation Time** | %s | %s | %s |\n", formatMS($stats['plain object']['creationTimeMs'] ?? 0), formatMS($stats['value object']['creationTimeMs'] ?? 0), formatMS($stats['value object minimal']['creationTimeMs'] ?? 0));
        $newSection .= sprintf("| **Memory Used (Heap)** | %s | %s | %s |\n", formatMB($stats['plain object']['memoryUsedMB'] ?? 0), formatMB($stats['value object']['memoryUsedMB'] ?? 0), formatMB($stats['value object minimal']['memoryUsedMB'] ?? 0));
        $newSection .= sprintf("| **Traversal Time** | %s | %s | %s |\n", formatMS($stats['plain object']['plainTraversalTimeMs'] ?? 0), formatMS($stats['value object']['plainTraversalTimeMs'] ?? 0), formatMS($stats['value object minimal']['plainTraversalTimeMs'] ?? 0));
        $newSection .= sprintf("| **Property Access Time** | %s | %s | %s |\n", formatMS($stats['plain object']['propAccessTimeMs'] ?? 0), formatMS($stats['value object']['propAccessTimeMs'] ?? 0), formatMS($stats['value object minimal']['propAccessTimeMs'] ?? 0));
        $newSection .= sprintf("| **Filtering Time** | %s | %s | %s |\n", formatMS($stats['plain object']['filterTimeMs'] ?? 0), formatMS($stats['value object']['filterTimeMs'] ?? 0), formatMS($stats['value object minimal']['filterTimeMs'] ?? 0));
        $newSection .= sprintf("| **Mutation Time** | %s | %s | %s |\n", formatMS($stats['plain object']['mutationTimeMs'] ?? 0), formatMS($stats['value object']['mutationTimeMs'] ?? 0), formatMS($stats['value object minimal']['mutationTimeMs'] ?? 0));
        $newSection .= sprintf("| **Delete Property Time** | %s | %s | %s |\n", formatMS($stats['plain object']['deletePropertyTimeMs'] ?? 0), formatMS($stats['value object']['deletePropertyTimeMs'] ?? 0), formatMS($stats['value object minimal']['deletePropertyTimeMs'] ?? 0));
    }

    if ($statsVar) {
        $newSection .= "\n### 2. Variable Properties (Polymorphic Shapes)\n\n";
        $newSection .= "| Metric | Plain Object (Array) | Value Object | Value Object Minimal |\n";
        $newSection .= "| :--- | :---: | :---: | :---: | :---: |\n";
        $newSection .= sprintf("| **Creation Time** | %s | %s | %s |\n", formatMS($statsVar['plain object']['creationTimeMs'] ?? 0), formatMS($statsVar['value object']['creationTimeMs'] ?? 0), formatMS($statsVar['value object minimal']['creationTimeMs'] ?? 0));
        $newSection .= sprintf("| **Memory Used (Heap)** | %s | %s | %s |\n", formatMB($statsVar['plain object']['memoryUsedMB'] ?? 0), formatMB($statsVar['value object']['memoryUsedMB'] ?? 0), formatMB($statsVar['value object minimal']['memoryUsedMB'] ?? 0));
        $newSection .= sprintf("| **Traversal Time** | %s | %s | %s |\n", formatMS($statsVar['plain object']['plainTraversalTimeMs'] ?? 0), formatMS($statsVar['value object']['plainTraversalTimeMs'] ?? 0), formatMS($statsVar['value object minimal']['plainTraversalTimeMs'] ?? 0));
        $newSection .= sprintf("| **Property Access Time** | %s | %s | %s |\n", formatMS($statsVar['plain object']['propAccessTimeMs'] ?? 0), formatMS($statsVar['value object']['propAccessTimeMs'] ?? 0), formatMS($statsVar['value object minimal']['propAccessTimeMs'] ?? 0));
        $newSection .= sprintf("| **Filtering Time** | %s | %s | %s |\n", formatMS($statsVar['plain object']['filterTimeMs'] ?? 0), formatMS($statsVar['value object']['filterTimeMs'] ?? 0), formatMS($statsVar['value object minimal']['filterTimeMs'] ?? 0));
        $newSection .= sprintf("| **Mutation Time** | %s | %s | %s |\n", formatMS($statsVar['plain object']['mutationTimeMs'] ?? 0), formatMS($statsVar['value object']['mutationTimeMs'] ?? 0), formatMS($statsVar['value object minimal']['mutationTimeMs'] ?? 0));
        $newSection .= sprintf("| **Delete Property Time** | %s | %s | %s |\n", formatMS($statsVar['plain object']['deletePropertyTimeMs'] ?? 0), formatMS($statsVar['value object']['deletePropertyTimeMs'] ?? 0), formatMS($statsVar['value object minimal']['deletePropertyTimeMs'] ?? 0));
    }

    $newSection .= "\n<!-- BENCHMARK_RESULTS_END -->";

    $readmeContent = "";
    if (file_exists($readmePath)) {
        $readmeContent = file_get_contents($readmePath);
    }

    $startMarker = "<!-- BENCHMARK_RESULTS_START -->";
    $endMarker = "<!-- BENCHMARK_RESULTS_END -->";

    if (strpos($readmeContent, $startMarker) !== false && strpos($readmeContent, $endMarker) !== false) {
        $startIndex = strpos($readmeContent, $startMarker);
        $endIndex = strpos($readmeContent, $endMarker) + strlen($endMarker);
        $readmeContent = substr($readmeContent, 0, $startIndex) . $newSection . substr($readmeContent, $endIndex);
    } else {
        // Look for ## Benchmark Results and replace or append
        $regex = '/## Benchmark Results \([\d,]+ Entries\)[\s\S]*?(?=## Key Findings)/';
        if (preg_match($regex, $readmeContent, $matches, PREG_OFFSET_CAPTURE)) {
            $readmeContent = preg_replace($regex, $newSection . "\n\n", $readmeContent);
        } else {
            // Append right before "## Key Findings" or at end
            $keyFindingsPos = strpos($readmeContent, "## Key Findings");
            if ($keyFindingsPos !== false) {
                $readmeContent = substr($readmeContent, 0, $keyFindingsPos) . $newSection . "\n\n" . substr($readmeContent, $keyFindingsPos);
            } else {
                $readmeContent .= "\n\n" . $newSection;
            }
        }
    }

    file_put_contents($readmePath, $readmeContent);
    echo "Successfully updated README.md from stats!\n";
}

run();

File: language_construct_performance/go/scripts/update_readme.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"strconv"
)

func formatMB(mb interface{}) string {
	if mb == nil {
		return "0 MB"
	}
	var val float64
	switch v := mb.(type) {
	case float64:
		val = v
	case int:
		val = float64(v)
	default:
		return fmt.Sprintf("%v MB", mb)
	}
	gb := val / 1024.0
	// Not adding comma formatting for simplicity here
	return fmt.Sprintf("%d MB (~%.1f GB)", int(val), gb)
}

func formatMS(ms interface{}) string {
	if ms == nil {
		return "0 ms"
	}
	var val int
	switch v := ms.(type) {
	case float64:
		val = int(v)
	case int:
		val = v
	default:
		return fmt.Sprintf("%v ms", ms)
	}
	// Simple comma formatting could be done, but keep it simple
	s := strconv.Itoa(val)
	n := len(s)
	if n <= 3 {
		return s + " ms"
	}
	var b strings.Builder
	for i, c := range s {
		if i > 0 && (n-i)%3 == 0 {
			b.WriteRune(',')
		}
		b.WriteRune(c)
	}
	return b.String() + " ms"
}

func main() {
	readmePath := "README.md"
	statsPath := filepath.Join("data", "stats.json")
	statsVarPath := filepath.Join("data", "stats_variable.json")

	if _, err := os.Stat(readmePath); os.IsNotExist(err) {
		fmt.Fprintln(os.Stderr, "README.md not found!")
		os.Exit(1)
	}

	var stats map[string]interface{}
	var statsVar map[string]interface{}

	if b, err := os.ReadFile(statsPath); err == nil {
		json.Unmarshal(b, &stats)
	}
	if b, err := os.ReadFile(statsVarPath); err == nil {
		json.Unmarshal(b, &statsVar)
	}

	if stats == nil && statsVar == nil {
		fmt.Fprintln(os.Stderr, "No valid stats files found in data/ directory!")
		os.Exit(1)
	}

	var numEntries float64 = 20000000
	if stats != nil && stats["numEntries"] != nil {
		numEntries = stats["numEntries"].(float64)
	} else if statsVar != nil && statsVar["numEntries"] != nil {
		numEntries = statsVar["numEntries"].(float64)
	}
	
	sNum := strconv.Itoa(int(numEntries))
	var formattedEntries string
	for i, c := range sNum {
		if i > 0 && (len(sNum)-i)%3 == 0 {
			formattedEntries += ","
		}
		formattedEntries += string(c)
	}

	newSection := "<!-- BENCHMARK_RESULTS_START -->\n"
	newSection += "## Benchmark Results (" + formattedEntries + " Entries)\n\n"
	newSection += "Here are the actual measured results from running the isolated benchmark suite under Go with **" + formattedEntries + " entries**:\n"

	if stats != nil {
		newSection += "\n### 1. Fixed Properties (Uniform Structural Shape)\n\n"
		newSection += "| Metric | Plain Object | Value Object | Value Object Minimal |\n"
		newSection += "| :--- | :---: | :---: | :---: | :---: |\n"
		
		po := stats["plain object"].(map[string]interface{})
		vo := stats["value object"].(map[string]interface{})
		vom := stats["value object minimal"].(map[string]interface{})

		newSection += fmt.Sprintf("| **Creation Time** | %s | %s | %s |\n", formatMS(po["creationTimeMs"]), formatMS(vo["creationTimeMs"]), formatMS(vom["creationTimeMs"]))
		newSection += fmt.Sprintf("| **Memory Used (Heap)** | %s | %s | %s |\n", formatMB(po["memoryUsedMB"]), formatMB(vo["memoryUsedMB"]), formatMB(vom["memoryUsedMB"]))
		newSection += fmt.Sprintf("| **Traversal Time** | %s | %s | %s |\n", formatMS(po["plainTraversalTimeMs"]), formatMS(vo["plainTraversalTimeMs"]), formatMS(vom["plainTraversalTimeMs"]))
		newSection += fmt.Sprintf("| **Property Access Time** | %s | %s | %s |\n", formatMS(po["propAccessTimeMs"]), formatMS(vo["propAccessTimeMs"]), formatMS(vom["propAccessTimeMs"]))
		newSection += fmt.Sprintf("| **Filtering Time** | %s | %s | %s |\n", formatMS(po["filterTimeMs"]), formatMS(vo["filterTimeMs"]), formatMS(vom["filterTimeMs"]))
		newSection += fmt.Sprintf("| **Mutation Time** | %s | %s | %s |\n", formatMS(po["mutationTimeMs"]), formatMS(vo["mutationTimeMs"]), formatMS(vom["mutationTimeMs"]))
		newSection += fmt.Sprintf("| **Delete Property Time** | %s | %s | %s |\n", formatMS(po["deletePropertyTimeMs"]), formatMS(vo["deletePropertyTimeMs"]), formatMS(vom["deletePropertyTimeMs"]))
	}

	if statsVar != nil {
		newSection += "\n### 2. Variable Properties (Polymorphic Shapes)\n\n"
		newSection += "| Metric | Plain Object | Value Object | Value Object Minimal |\n"
		newSection += "| :--- | :---: | :---: | :---: | :---: |\n"
		
		po := statsVar["plain object"].(map[string]interface{})
		vo := statsVar["value object"].(map[string]interface{})
		vom := statsVar["value object minimal"].(map[string]interface{})

		newSection += fmt.Sprintf("| **Creation Time** | %s | %s | %s |\n", formatMS(po["creationTimeMs"]), formatMS(vo["creationTimeMs"]), formatMS(vom["creationTimeMs"]))
		newSection += fmt.Sprintf("| **Memory Used (Heap)** | %s | %s | %s |\n", formatMB(po["memoryUsedMB"]), formatMB(vo["memoryUsedMB"]), formatMB(vom["memoryUsedMB"]))
		newSection += fmt.Sprintf("| **Traversal Time** | %s | %s | %s |\n", formatMS(po["plainTraversalTimeMs"]), formatMS(vo["plainTraversalTimeMs"]), formatMS(vom["plainTraversalTimeMs"]))
		newSection += fmt.Sprintf("| **Property Access Time** | %s | %s | %s |\n", formatMS(po["propAccessTimeMs"]), formatMS(vo["propAccessTimeMs"]), formatMS(vom["propAccessTimeMs"]))
		newSection += fmt.Sprintf("| **Filtering Time** | %s | %s | %s |\n", formatMS(po["filterTimeMs"]), formatMS(vo["filterTimeMs"]), formatMS(vom["filterTimeMs"]))
		newSection += fmt.Sprintf("| **Mutation Time** | %s | %s | %s |\n", formatMS(po["mutationTimeMs"]), formatMS(vo["mutationTimeMs"]), formatMS(vom["mutationTimeMs"]))
		newSection += fmt.Sprintf("| **Delete Property Time** | %s | %s | %s |\n", formatMS(po["deletePropertyTimeMs"]), formatMS(vo["deletePropertyTimeMs"]), formatMS(vom["deletePropertyTimeMs"]))
	}

	newSection += "\n<!-- BENCHMARK_RESULTS_END -->"

	b, _ := os.ReadFile(readmePath)
	readmeContent := string(b)

	startMarker := "<!-- BENCHMARK_RESULTS_START -->"
	endMarker := "<!-- BENCHMARK_RESULTS_END -->"

	if strings.Contains(readmeContent, startMarker) && strings.Contains(readmeContent, endMarker) {
		startIdx := strings.Index(readmeContent, startMarker)
		endIdx := strings.Index(readmeContent, endMarker) + len(endMarker)
		readmeContent = readmeContent[:startIdx] + newSection + readmeContent[endIdx:]
	} else {
		// Just append it
		readmeContent += "\n\n" + newSection
	}

	os.WriteFile(readmePath, []byte(readmeContent), 0644)
	fmt.Println("Successfully updated README.md from stats!")
}
